<a href="https://colab.research.google.com/github/DeepuBenson/Python/blob/main/Day1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Python: A multi paradigm programming language**

A few examples for functions in Python

In [2]:
def intsum(n):
    return sum(range(n+1))

print(intsum(100))

5050


In [3]:
def foo(n):
    print(n+n)

In [4]:
foo('1234')
foo(1234)
foo([1,2,3,4])
foo((1,2,3,4))

12341234
2468
[1, 2, 3, 4, 1, 2, 3, 4]
(1, 2, 3, 4, 1, 2, 3, 4)


In [5]:
n=int(input('Enter a positive number: '))
if n > 1:
    for i in range(2,n-1):
        if n % i == 0:
            print(n, "is composite")
            break
    else:     #Python for loops can also have else!                      
        print(n, "is prime")
else:         
    print(n, "is neither prime nor composite")

Enter a positive number: 123
123 is composite


In [6]:
def intsum(n):
    if n==1:
        return 1
    else:
        return n+intsum(n-1)

x=int(input('Enter a number : '))
print(intsum(x))

Enter a number : 122
7503


In [7]:
def fact(n):
    if n==1:
        return 1
    else:
        return n*fact(n-1)

x=int(input('Enter a number : '))
print(fact(x))

Enter a number : 6
720


A few examples for classes in Python

In [8]:
class ClassName:
    #Body of the class
    #should not be empty
    pass 

In [11]:
class OurClass:
    x=123

oc1=OurClass() #First Object created
oc2=OurClass() #Second Object created

print(oc1.x)
print(oc2.x)

oc1.x=456

print(oc1.x)   # ???
print(oc2.x)   # ???

print(id(oc1)) # ???
print(id(oc2)) # ???

123
123
456
123
140611706230624
140611706230432


In [12]:
class Check:
    a=100
    def __init__(self,n): #Similar to a constructor in C++/Java
        self.n = n
    def checking(self):   # This is called a method
        print("---->",Check.a)
        if self.n % 2 == 0:
            print(self.n,'is an even number')
        else:
            print(self.n,'is an odd number')

    def ch(self):
        Check.a=333

x=Check(123)
y=Check(456)
x.ch()
x.checking()
y.checking()

----> 333
123 is an odd number
----> 333
456 is an even number


Functional Programming with Python

In [13]:
lam = lambda n: n+n

In [14]:
lam(123)

246

In [15]:
lam('123')

'123123'

In [16]:
def foo(n):
    return lambda x: x**n

In [17]:
square = foo(2)

In [18]:
print(square)

<function foo.<locals>.<lambda> at 0x7fe2b9f23b80>


In [19]:
square(5)

25

In [20]:
cube = foo(3)

In [21]:
cube(5)

125

In [25]:
x='123'
print(int(x))

123


In [40]:
y=['11', '22', '33', '44', '55']

In [41]:
print(int(y))   #Intentional Mistake!

TypeError: ignored

In [42]:
print(list(map(int,y)))

[11, 22, 33, 44, 55]


In [45]:
import math as m
y = [(x*m.pi)/180 for x in range(0, 361,30)] # Example for List Comprehension. A very useful feature.
print(y)

[0.0, 0.5235987755982988, 1.0471975511965976, 1.5707963267948966, 2.0943951023931953, 2.6179938779914944, 3.141592653589793, 3.6651914291880923, 4.1887902047863905, 4.71238898038469, 5.235987755982989, 5.759586531581287, 6.283185307179586]


In [36]:
print(list(map(m.sin, y)))

[0.0, 0.49999999999999994, 0.8660254037844386, 1.0, 0.8660254037844387, 0.49999999999999994, 1.2246467991473532e-16, -0.5000000000000001, -0.8660254037844384, -1.0, -0.8660254037844386, -0.5000000000000004, -2.4492935982947064e-16]


**Why Sin(30) is shown as 0.49999999999999994 instead of 0.5?**

A long, detailed and difficult explanation can be found here ---> [What Every Computer Scientist Should Know About Floating-Point Arithmetic](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html)

**Optional reading material**

In [46]:
print(m.sqrt(3)/2)

0.8660254037844386


In [23]:
def foo(n):
    return n+n

In [ ]:
x=[1, 2, 3, 4, 5]